### PR TITLE
move sallyport and rcrt1 into enarx-keep

### PR DESCRIPTION
**Pull request by haraldh**
- also move the shims in the `internal` directory
- extend the `build.rs` to cope with `Cargo.tml` -> `Cargo.toml`
  renaming

### CODE REVIEW

**Review by mbestavros — DISMISSED** [review]

File: enarx-keep/build.rs [hunk 1]
```diff
@@ -59,22 +59,42 @@ fn main() {
...
+    for entry in std::fs::read_dir("internal").unwrap() {
+        let shim_path = entry.unwrap().path();
+
+        let cargo_toml = shim_path.join("Cargo.toml");
+        let cargo_tml = shim_path.join("Cargo.tml");
+
+        if cargo_tml.exists() {
+            std::fs::copy(cargo_tml, cargo_toml).unwrap();
+        }
+    }
```
Comment (npmccallum) [inline enarx-keep/build.rs:74]: It is my understanding that this modifies files in the source directory. I think that's frowned upon. But... they can continue frowning until `cargo` provides us a better solution.

File: enarx-keep/build.rs [hunk 2]
```diff
@@ -59,22 +59,42 @@ fn main() {
         })
         .collect();
 
-    for entry in std::fs::read_dir("shims").unwrap() {
+    // internal crates are not included, if there is a `Cargo.toml` file
+    // trick cargo by renaming the `Cargo.toml` to `Cargo.tml` before
+    // publishing and rename it back here.
+    for entry in std::fs::read_dir("internal").unwrap() {
+        let shim_path = entry.unwrap().path();
```
Comment (npmccallum) [inline enarx-keep/build.rs:66]:
```suggestion
        let path = entry.unwrap().path();
```

File: enarx-keep/src/backend/kvm/vm/cpu.rs [hunk 3]
```diff
@@ -7,11 +7,11 @@ use crate::backend::kvm::shim::{
     MemInfo, SYSCALL_TRIGGER_PORT, SYS_ENARX_BALLOON_MEMORY, SYS_ENARX_MEM_INFO,
 };
 use crate::backend::{Command, Thread};
+use crate::sallyport;
```
Comment (npmccallum) [inline enarx-keep/src/backend/kvm/vm/cpu.rs:10]: I'd prefer to stick with `use crate::sallyport::Reply`... Is there some reason for this change?

File: enarx-keep/src/backend/mod.rs [hunk 4]
```diff
@@ -7,12 +7,12 @@ pub mod sgx;
 mod probe;
 
 use crate::binary::Component;
+use crate::sallyport;
```
Comment (npmccallum) [inline enarx-keep/src/backend/mod.rs:10]: `sallyport::Block`

File: enarx-keep/src/main.rs [hunk 5]
```diff
@@ -4,9 +4,14 @@
...
 #![deny(clippy::all)]
 #![deny(missing_docs)]
+#![feature(asm)]
 
 mod backend;
 mod binary;
+mod sallyport;
+
+// workaround for sallyport tests, until we have internal crates
+pub use sallyport::Request;
```
Comment (npmccallum) [inline enarx-keep/src/main.rs:14]: Why not `pub mod sallyport`?

File: enarx-keep/build.rs (hunk 1 above)
Comment (haraldh) [reply enarx-keep/build.rs:74]: yep

File: enarx-keep/build.rs (hunk 2 above)
Comment (haraldh) [reply enarx-keep/build.rs:66]: what's your problem here?

File: enarx-keep/src/main.rs (hunk 5 above)
Comment (haraldh) [reply enarx-keep/src/main.rs:14]: doesn't make a difference. The macro in the tests wants `$crate::Request`

File: enarx-keep/src/backend/kvm/vm/cpu.rs (hunk 3 above)
Comment (haraldh) [reply enarx-keep/src/backend/kvm/vm/cpu.rs:10]: no reason, changed `sallyport::Block` instead then

File: enarx-keep/src/backend/mod.rs (hunk 4 above)
Comment (haraldh) [reply enarx-keep/src/backend/mod.rs:10]: ok

File: enarx-keep/build.rs (hunk 2 above)
Comment (haraldh) [reply enarx-keep/build.rs:66]: ah, oh, right

**Comment by lkatalin** [comment]
Looks like this resolves #856

**Pull request merged**